enhancements/installer/vsphere-ipi-multiple-datacenter-clusters.md

```diff
@@ -0,0 +1,474 @@
+api-approvers:
+  - ""
+creation-date: 2021-09-21
+last-updated: 2021-10-26
+status: implementable
+see-also:
+  - "/enhancements/"
+replaces:
+  -
+superseded-by:
+  -
+---
+
+# Support Multiple vSphere vCenter Datacenters and Clusters in IPI
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The goal of this enhancement is to provide the ability to install in a
+vSphere environment with multiple datacenters and clusters.
+
+This will be an opinionated design, the vCenter datacenter will always be a `region`
+and a vCenter cluster will always be a `zone`.
+
+## Motivation
+
+Users of OpenShift would like the ability to deploy
+within multiple datacenters and clusters to increase
+reliability. Customers would also like to take advantage
+of the concept of regions and zoning that this type of deployment would
+offer.
+
+- https://issues.redhat.com/browse/RFE-845
+- https://issues.redhat.com/browse/OCPPLAN-4927
+
+### Goals
+
+- Support regions and zones in vSphere using multiple datacenters (region) and
+clusters (zone)
+- Support installation into multiple datacenters and multiple clusters
+
+### Non-Goals
+
+- Support for VMware vSphere in-tree or out-of-tree storage provider
+- Support multiple subnets
+- Support multiple vcenters
+
+Note: The platform spec will be modified to support this at a future date.
+Only a single item in `vcenters` will be supported for the initial release.
+
+## Proposal
+
+Modification of the installer to support the provisioning of masters in defined
+datacenters and clusters.
+
+### installer
+
+#### Platform spec
+
+The platform spec needs to be modified to support our initial goals of multiple
+datacenters (regions) and clusters (zones) and vcenters.
+
+We are adding one additional parameter to the Platform struct:
+
+- `VCenters`
+
+`VCenters` will contain the connection details for each individual vcenter
+and the `Region` associated with that `VCenter`. The
+existing platform spec vcenter parameters will be deprecated but _not_ removed
+or remove support for using those parameters.
+
+Regions will contain the datacenter and `Zones` that its associated with.
+Each region's zones will be associated with the
+additional configuration data required to provision virtual machines within
+vSphere. Including `Cluster`, `Network` (port group), `ResourcePool` and the
+`Datastore`.
+
+```golang
+type Platform struct {
+...
+  VCenters []VCenter `json:"vcenters,omitempty"`
+}
+
+type VCenter struct {
+  Server string `json:"server,omitempty"`
+  Port uint `json:"port,omitempty"`
+  User string `json:"user,omitempty"`
+  Password string `json:"password,omitempty"`
```

Comment:
Credentials in plaintext? I know this is required for terraform, is it needed anywhere else? Not in the CCM right as that reads from a secret?

Reply:
> Not in the CCM right as that reads from a secret? 
Correct, CCM reads from a secret.

Reply:
It might be helpful to include comments within this code block to explain each of the fields and in particular, for fields like this when they are used and when they are no longer needed (eg these aren't put into the cluster at any point - unless they are?)

Reply:
added